### ISSUE TITLE
conda needs to be more forgiving of bad channel responses

### ISSUE THREAD

**Issue by ijstokes** [conda/conda#2387]
While investigating the cause of #1036 I looked into the `.condarc` file which specifies:

```
channels:
  - https://conda.anaconda.org/t/an-c89909f5-b1ad-4c73-9500-9e5f0e710c33/anaconda-nb-extensions
  - http://localhost:8080/conda/anaconda
  - defaults
```

The second entry is for a local deployment of Anaconda Repository.  It turns out the license had expired, so the responses conda got from that channel were bogus.  That is mentioned in more than one place, but at least one of them is https://github.com/conda/conda/issues/1189.

Because Anaconda Repository did not return the JSON result that conda was expecting, conda then threw this stack trace:

```
$ sudo -u wakari /opt/wakari/miniconda/bin/conda create -p /tmp/testenv python
Fetching package metadata: ....An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:
  
    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/opt/wakari/miniconda/bin/conda", line 5, in <module>
    sys.exit(main())
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/cli/main.py", line 195, in main
    args_func(args, p)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/cli/main.py", line 202, in args_func
    args.func(args, p)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/cli/main_create.py", line 49, in execute
    install.install(args, parser, 'create')
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/cli/install.py", line 239, in install
    offline=args.offline)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/cli/common.py", line 597, in get_index_trap
    return get_index(*args, **kwargs)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/api.py", line 42, in get_index
    unknown=unknown)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/utils.py", line 118, in __call__
    value = self.func(*args, **kw)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/fetch.py", line 264, in fetch_index
    reversed(channel_urls))
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/fetch.py", line 263, in <lambda>
    use_cache=use_cache, session=session)),
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/fetch.py", line 65, in func
    res = f(*args, **kwargs)
  File "/opt/wakari/miniconda/lib/python2.7/site-packages/conda/fetch.py", line 103, in fetch_repodata
    cache = json.loads(bz2.decompress(resp.content).decode('utf-8'))
IOError: invalid data stream
```

Because this kind of scenario is possible for several reasons I believe conda needs to be made more robust against getting back bogus JSON.

Of course Anaconda Repository should also be more helpful and not pretend to return `repodata.json` (which I think is happening here), so I will also submit an issue there.

**Comment by kalefranz** [comment]
Going to go ahead and call this a bug.

> It turns out the license had expired, so the responses conda got from that channel were bogus.

What's the actual response from the anaconda repo server when this happens?  Is it actual json or some nasty html with a 40X error?

**Comment by ijstokes** [comment]
@kalefranz here is the output with `--debug` enabled.  My reading of it is that Anaconda Repository wrongly returns a 200 (which is the issue I've raised in the associated Anaconda Repo issue referenced above).

```
Fetching package metadata: INFO:stdoutlog:Fetching package metadata: 
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): repo.continuum.io
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/pro/noarch/repodata.json.bz2 HTTP/1.1" 304 0
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/pro/noarch/',) kwargs {'session': <conda.connection.CondaSession object at 0x7f0a7846b150>, 'use_cache': False}
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/pro/linux-64/repodata.json.bz2 HTTP/1.1" 304 0
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/pro/linux-64/',) kwargs {'session': <conda.connection.CondaSession object at 0x7f0a7846b150>, 'use_cache': False}
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/free/noarch/repodata.json.bz2 HTTP/1.1" 304 0
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/free/noarch/',) kwargs {'session': <conda.connection.CondaSession object at 0x7f0a7846b150>, 'use_cache': False}
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/free/linux-64/repodata.json.bz2 HTTP/1.1" 304 0
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/free/linux-64/',) kwargs {'session': <conda.connection.CondaSession object at 0x7f0a7846b150>, 'use_cache': False}
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): localhost
DEBUG:requests.packages.urllib3.connectionpool:"GET /conda/anaconda/noarch/repodata.json.bz2 HTTP/1.1" 302 343
INFO:requests.packages.urllib3.connectionpool:Resetting dropped connection: localhost
DEBUG:requests.packages.urllib3.connectionpool:"GET /admin/license?next=%2Fconda%2Fanaconda%2Fnoarch%2Frepodata.json.bz2 HTTP/1.1" 200 3351
An unexpected error has occurred, please consider sending the
```

**Comment by kalefranz** [comment]
What is the content/body of the HTML response though?  That's what I was more interested in...

**Comment by ijstokes** [comment]
Sorry, no idea. Would be a bit of work for me to reproduce, but you could do it with a Trial VM instance by following the steps above and just make sure the license file is not present for Anaconda Repository.

**Comment by kalefranz** [comment]
I think that has already been fixed in 4.1.x now.
